### PR TITLE
HomeWork_10

### PR DESCRIPTION
**Pull request by Sofasmile**

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -1,7 +1,4 @@
 *.iml
-*.xml
 .idea
 *.patch
```
Comment (Anderwerty) [inline .gitignore]: Why did you add it into gitignore?

File: listStructure.xml [hunk 2]
```diff
@@ -0,0 +1,17 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (Anderwerty) [inline listStructure.xml]: You should not commit this file.

File: src/main/java/com/expression/RegularUtil.java [hunk 3]
```diff
@@ -0,0 +1,60 @@
+package com.expression;
+
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+public class RegularUtil {
+    private static final String phoneNumber = "^\\+\\d{2}\\(\\d{3}\\)\\d{3}\\-\\d{2}\\-\\d{2}$";
```
Comment (Anderwerty) [inline src/main/java/com/expression/RegularUtil.java]: Naming is not right.
https://mate-academy.github.io/style-guides/java/java.html#s5.2.4-constant-names

File: src/main/java/com/expression/RegularUtil.java [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+        } else {
+            return false;
+        }
+    }
+
+    private static boolean isCorrectDate(String dd, String mm, String yyyy) {
+        int year = Integer.parseInt(yyyy);
+        if (dd.equals("31") && isCorrectMonth(mm)) {
+            return false;
+        } else if (dd.equals("30") && mm.equals("02")) {
```
Comment (Anderwerty) [inline src/main/java/com/expression/RegularUtil.java]: you could remove a word `else`, as in previous if block you have a return statement .

File: src/main/java/com/expression/RegularUtil.java [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+            return false;
+        } else if (dd.equals("29") && mm.equals("02") && (year % 4 != 0)) {
```
Comment (Anderwerty) [inline src/main/java/com/expression/RegularUtil.java]: the same. Also
 
Every year that is exactly divisible by four is a leap year, except for years that are exactly divisible by 100, but these centurial years are leap years if they are exactly divisible by 400. For example, the years 1700, 1800, and 1900 are not leap years, but the years 1600 and 2000 are.(https://en.wikipedia.org/wiki/Leap_year)

File: src/main/java/com/expression/RegularUtil.java [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+            return false;
+        } else {
+            return true;
+        }
+    }
+
+    private static boolean isCorrectMonth(String month) {
+        return (month.equals("02") | month.equals("04")
```
Comment (Anderwerty) [inline src/main/java/com/expression/RegularUtil.java:55]:
```
(month.equals("02") | month.equals("04")
                | month.equals("06") | month.equals("09")
                | month.equals("11")) 
```
This statment already return bolean result. So  `? true : false` could be removed.

File: src/main/java/com/xml/XmlGenetare.java [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+    private void makeXmlFile(List list) {
+        FileOutputStream xmlFile = null;
+        try {
+            xmlFile = new FileOutputStream("listStructure.xml");
+        } catch (FileNotFoundException e) {
+            e.getMessage();
+        }
+        java.beans.XMLEncoder xmlEncoder = new java.beans.XMLEncoder(xmlFile);
+        xmlEncoder.writeObject(list);
+        xmlEncoder.close();
```
Comment (Anderwerty) [inline src/main/java/com/xml/XmlGenetare.java]: Could we use try-with-resources or finally block to close streams?

File: src/main/java/com/xml/XmlGenetare.java [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+    }
+
+    private void makeXmlFile(List list) {
+        FileOutputStream xmlFile = null;
+        try {
+            xmlFile = new FileOutputStream("listStructure.xml");
+        } catch (FileNotFoundException e) {
+            e.getMessage();
+        }
+        java.beans.XMLEncoder xmlEncoder = new java.beans.XMLEncoder(xmlFile);
```
Comment (Anderwerty) [inline src/main/java/com/xml/XmlGenetare.java]: Why don't you import this class?

File: src/main/java/com/xml/XmlGenetare.java [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+
+        makeXmlFile(listStructure);
+    }
+
+    private void makeXmlFile(List list) {
+        FileOutputStream xmlFile = null;
+        try {
+            xmlFile = new FileOutputStream("listStructure.xml");
+        } catch (FileNotFoundException e) {
+            e.getMessage();
```
Comment (Anderwerty) [inline src/main/java/com/xml/XmlGenetare.java]: Please use logger.

File: src/test/java/com/expression/RegularUtilTest.java [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+import org.junit.Before;
+import org.junit.Test;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertTrue;
+
+public class RegularUtilTest {
```
Comment (Anderwerty) [inline src/test/java/com/expression/RegularUtilTest.java:12]: Optional: look at parametrized junit tests.

File: src/test/java/com/xml/XmlGenetareTest.java [hunk 11]
```diff
@@ -0,0 +1,11 @@
+package com.xml;
+
+import org.junit.Test;
+
+public class XmlGenetareTest {
+    @Test
+    public void testMakeMultiLayerStructure() {
+        XmlGenetare xmlGenetare = new XmlGenetare();
```
Comment (Anderwerty) [inline src/test/java/com/xml/XmlGenetareTest.java:8]: This test should have assert.

**Review by Anderwerty — CHANGES_REQUESTED** [review]

File: .gitignore (hunk 1 above)
Comment (Sofasmile) [reply .gitignore]: These files came up when I solved the problem with my repository. So now, I don`t know if I can remove them

File: src/test/java/com/xml/XmlGenetareTest.java (hunk 11 above)
Comment (Sofasmile) [reply src/test/java/com/xml/XmlGenetareTest.java:8]: I don't know how to write a test correctly
Maybe you will help me?

File: .gitignore (hunk 1 above)
Comment (Anderwerty) [reply .gitignore]: I think you could delete *.patch form gitignore

File: .gitignore [hunk 12]
```diff
@@ -1,7 +1,6 @@
...
 target
-
+*.xml
```
Comment (Anderwerty) [inline .gitignore]: Why it is here? It should not be here. Remove, please.

File: src/test/java/com/xml/XmlGenetareTest.java (hunk 11 above)
Comment (Anderwerty) [reply src/test/java/com/xml/XmlGenetareTest.java:8]: I see Alexander answer in chat. You don't need to use assert.

File: src/main/java/com/xml/XmlGenetare.java [hunk 13]
```diff
@@ -0,0 +1,37 @@
...
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.HashMap;
+import java.beans.XMLEncoder;
+
+public class XmlGenetare {
+    private static final Logger LOGGER = Logger.getLogger(XmlGenetare.class.getName());
```
Comment (Anderwerty) [inline src/main/java/com/xml/XmlGenetare.java:16]: Please add one blank line between constant and method.

File: pom.xml [hunk 14]
```diff
@@ -0,0 +1,36 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<project xmlns="http://maven.apache.org/POM/4.0.0"
+         xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+    <modelVersion>4.0.0</modelVersion>
+
+    <groupId>com.collection</groupId>
+    <artifactId>collection-test</artifactId>
+    <version>1.0-SNAPSHOT</version>
```
Comment (Anderwerty) [inline pom.xml:9]: Optional: add a blank line between lines 9 and 10.

File: pom.xml [hunk 15]
```diff
@@ -0,0 +1,36 @@
...
+    <build>
+        <plugins>
+            <plugin>
+                <groupId>org.apache.maven.plugins</groupId>
+                <artifactId>maven-compiler-plugin</artifactId>
+                <configuration>
+                    <source>7</source>
```
Comment (Anderwerty) [inline pom.xml]: It should be 1.8 or 8 but not 7. Am I right?
The same for target

**Review by Anderwerty — CHANGES_REQUESTED** [review]
Please fix

**Review by Anderwerty — APPROVED** [review]
